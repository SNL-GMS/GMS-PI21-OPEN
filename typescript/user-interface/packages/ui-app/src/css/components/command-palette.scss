.command-palette-overlay__contents {
  width: 33vw;
  max-width: 40rem;
  padding: 0.5em;
}
.command-palette {
  &__input {
    flex-grow: 1;
    border: 1px solid transparent;
    background: var(--gms-recessed, $gms-recessed);
    color: var(--gms-main, $gms-main);

    & input {
      font-size: 1rem;
    }
  }

  &__results-wrapper {
    max-height: max(50vh, 25em);
    overflow: auto;
    margin: 0 calc(-0.5em + 1px);
  }

  &__result-list {
    padding: 0;
    margin: 0;
    display: flex;
    flex-direction: column;
  }

  &__result {
    flex-grow: 1;
    background: var(--gms-prominent-background, $gms-prominent-background);
    padding: 0.5em;
    margin: 1px 0;
    list-style-type: none;

    &:first-of-type {
      margin-top: 0.5em;
    }

    &:hover {
      @include brighten;
    }

    &--selected {
      background-color: var(--gms-table-selection, $gms-table-selection);
      color: var(--table-selection-text-color, white);
    }
  }
  & .is-highlighted {
    background-color: rgba(var(--gms-main, $gms-main), 0.15);
  }
}
