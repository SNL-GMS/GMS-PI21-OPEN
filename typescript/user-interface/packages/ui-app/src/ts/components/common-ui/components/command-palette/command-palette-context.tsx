import * as React from 'react';

import type { Command, CommandScope } from './types';

/**
 * Stores data and functions used by commands
 */
export interface CommandPaletteContextData {
  commandPaletteIsVisible: boolean;
  registerCommands(commandsToRegister: Command[], scope: CommandScope);
  setCommandPaletteVisibility(visibility: boolean): void;
}

export const CommandPaletteContext: React.Context<CommandPaletteContextData> = React.createContext<
  CommandPaletteContextData
>(undefined);
