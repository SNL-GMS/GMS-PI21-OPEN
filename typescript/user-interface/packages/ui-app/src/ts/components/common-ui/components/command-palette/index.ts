export { ReduxCommandPaletteContainer as CommandPalette } from './command-palette-container';
export { CommandPaletteContext } from './command-palette-context';
export { CommandPaletteOverlay } from './command-palette-overlay';
export * from './types';
