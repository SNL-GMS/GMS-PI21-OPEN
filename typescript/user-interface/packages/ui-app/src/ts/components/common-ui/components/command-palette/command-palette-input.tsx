import { InputGroup } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { classList, useRestoreFocus } from '@gms/ui-util';
import React from 'react';

import { CommandPaletteContext } from './command-palette-context';
import type { CommandPaletteInputProps } from './types';

/**
 * Focuses on the command palette when it is loaded.
 * Returns focus to whatever had it when the CP is closed.
 * Cleans up the input field on close.
 *
 * @param onClearFocus a function that is called on close
 */
const useFocusManager: (
  onClearFocus: () => void
) => React.MutableRefObject<HTMLInputElement> = onClearFocus => {
  const context = React.useContext(CommandPaletteContext);
  const inputRef = React.useRef<HTMLInputElement>(null);
  useRestoreFocus(context.commandPaletteIsVisible);
  React.useEffect(() => {
    if (context.commandPaletteIsVisible) {
      inputRef.current.focus();
    } else {
      onClearFocus();
    }
    // !FIX ESLINT Validate and check REACT HOOK dependencies
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [context.commandPaletteIsVisible]);
  return inputRef;
};

/**
 * The form input and its keyboard listeners.
 */
export function CommandPaletteInput(props: CommandPaletteInputProps) {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { setSearchTerm, getSearchTerm, selectionManager } = props;
  const clearSearchTermOnClose = () => setSearchTerm('');

  const context = React.useContext(CommandPaletteContext);
  const inputRef = useFocusManager(clearSearchTermOnClose);

  return (
    <InputGroup
      inputRef={ref => {
        inputRef.current = ref;
      }}
      autoComplete="off"
      autoFocus
      className={classList({ 'command-palette__input': true })}
      leftIcon={IconNames.CHEVRON_RIGHT}
      name="command-palette"
      dir="ltr"
      placeholder="Command"
      value={getSearchTerm() ?? ''}
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          e.preventDefault();
          setSearchTerm('');
          context.setCommandPaletteVisibility(false);
        }
        if (e.key === 'ArrowUp') {
          e.preventDefault();
          selectionManager.selectPrevious();
        }
        if (e.key === 'ArrowDown') {
          e.preventDefault();
          selectionManager.selectNext();
        }
      }}
      onChange={(e: React.FormEvent<HTMLInputElement>) => {
        setSearchTerm(e.currentTarget.value);
        e.preventDefault();
      }}
    />
  );
}
