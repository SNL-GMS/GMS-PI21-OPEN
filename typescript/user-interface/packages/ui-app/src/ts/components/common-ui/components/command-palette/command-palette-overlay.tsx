import { OverlayWrapper } from '@gms/ui-core-components';
import * as React from 'react';

import { CommandPalette } from './command-palette';
import { CommandPaletteContext } from './command-palette-context';
import type { CommandPaletteOverlayProps } from './types';
import { hasCommandListChanged } from './utils';

/**
 * Wraps the command palette in an overlay, and manages that state.
 */
function BaseCommandPaletteOverlay(props: CommandPaletteOverlayProps) {
  const { showCommandPalette, commandActions } = props;
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { setCommandPaletteVisibility } = React.useContext(CommandPaletteContext);
  return (
    <OverlayWrapper
      isOpen={showCommandPalette}
      onClose={() => {
        setCommandPaletteVisibility(false);
      }}
    >
      <CommandPalette
        isVisible={showCommandPalette}
        commandActions={commandActions}
        defaultSearchTerms={['open display', 'close display']}
      />
    </OverlayWrapper>
  );
}

/**
 * Memoized so that it only updates if the command list has changed.
 * Compares CommandType and DisplayText. If these are unchanged, the
 * command palette will not update.
 */
export const CommandPaletteOverlay = React.memo(
  BaseCommandPaletteOverlay,
  (prevProps, nextProps) =>
    nextProps.showCommandPalette &&
    hasCommandListChanged(prevProps.commandActions, nextProps.commandActions)
);
