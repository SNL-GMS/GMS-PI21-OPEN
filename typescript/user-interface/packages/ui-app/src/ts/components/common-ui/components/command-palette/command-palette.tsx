/* eslint-disable react/destructuring-assignment */
import isEqual from 'lodash/isEqual';
import * as React from 'react';

import type { CommandPaletteContextData } from './command-palette-context';
import { CommandPaletteContext } from './command-palette-context';
import { CommandPaletteSearchManager } from './command-palette-search-manager';
import type { CommandPaletteProps, CommandPaletteState } from './types';
import { searchCommandsForTerms } from './utils';

/**
 * Command Palette creates a text input that searches a configured list of commands and presents
 * a dropdown of any commands that match the search.
 */
export class CommandPalette extends React.Component<CommandPaletteProps, CommandPaletteState> {
  /** Type for Command Palette context - provides the functions and data needed to create the commands */
  // eslint-disable-next-line react/static-property-placement
  public static contextType: React.Context<CommandPaletteContextData> = CommandPaletteContext;

  /** The Command Palette context - provides the functions and data needed to create the commands */
  // eslint-disable-next-line react/static-property-placement
  public declare context: React.ContextType<typeof CommandPaletteContext>;

  /** The previous context, used to check if the context has changed. */
  private prevContext: React.ContextType<typeof CommandPaletteContext>;

  public constructor(props) {
    super(props);
    this.state = {
      defaultSearchResults: props.defaultSearchResults ?? []
    };
  }

  /**
   * only update if the command palette is visible. Otherwise, it's a waste of resources.
   *
   * @returns boolean
   */
  public shouldComponentUpdate(nextProps: CommandPaletteProps): boolean {
    if (nextProps.isVisible) {
      return true;
    }
    return false;
  }

  /** If the context has changed, regenerate commands and default search results */
  public componentDidUpdate(): void {
    if (!isEqual(this.prevContext, this.context)) {
      this.prevContext = this.context;
      const updatedDefaultSearchResults = searchCommandsForTerms(
        this.props.defaultSearchTerms ?? [],
        this.props.commandActions
      );
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        defaultSearchResults: updatedDefaultSearchResults
      });
    }
  }

  public render(): JSX.Element {
    return (
      <CommandPaletteSearchManager
        commandActions={this.props.commandActions}
        defaultSearchResults={this.state.defaultSearchResults}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...this.props}
      />
    );
  }
}
