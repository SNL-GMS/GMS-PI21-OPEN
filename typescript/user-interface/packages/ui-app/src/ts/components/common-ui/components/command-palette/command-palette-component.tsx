/* eslint-disable react/prop-types */
/* eslint-disable react/destructuring-assignment */
import Immutable from 'immutable';
import * as React from 'react';

import type { CommandPaletteContextData } from './command-palette-context';
import { CommandPaletteContext } from './command-palette-context';
import { CommandPaletteOverlay } from './command-palette-overlay';
import type { Command, CommandPaletteComponentProps, CommandScope } from './types';

/**
 * Takes data and functions passed in from redux and updates the command palette context, which
 * is used by the command palette.
 * Adds the command palette overlay as a sibling after this component's children
 */
export const CommandPaletteComponent: React.FunctionComponent<React.PropsWithChildren<
  CommandPaletteComponentProps
  // eslint-disable-next-line react/function-component-definition
>> = ({
  children,
  commandPaletteIsVisible,
  setCommandPaletteVisibility
}: React.PropsWithChildren<CommandPaletteComponentProps>) => {
  const [commandActions, setCommandActions] = React.useState<
    Immutable.Map<CommandScope, Command[]>
  >(Immutable.Map());

  /**
   * Updates the commandAction map. Add or overwrite the list of commands keyed on the CommandScope.
   * Note, you can delete commands by registering an empty array for a given scope.
   *
   * @param commandsToRegister a list of command actions to register
   * @param scope the key to use to register these commands.
   */
  const registerCommands = (commandsToRegister: Command[], scope: CommandScope) => {
    setCommandActions(prevActions => prevActions.set(scope, commandsToRegister));
  };

  const commandPaletteContextData: CommandPaletteContextData = React.useMemo(
    () => ({
      commandPaletteIsVisible,
      registerCommands,
      setCommandPaletteVisibility
    }),
    [commandPaletteIsVisible, setCommandPaletteVisibility]
  );

  return (
    <CommandPaletteContext.Provider value={commandPaletteContextData}>
      {children}
      <CommandPaletteOverlay
        commandActions={commandActions.reduce(
          (allActions, actionList) => [...allActions, ...actionList],
          []
        )}
        showCommandPalette={commandPaletteIsVisible}
      />
    </CommandPaletteContext.Provider>
  );
};
