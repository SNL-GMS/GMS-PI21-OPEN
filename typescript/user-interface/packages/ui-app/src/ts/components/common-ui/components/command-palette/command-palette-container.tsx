import { compose } from '@gms/common-util';
import type { AppState } from '@gms/ui-state';
import { commonActions, setCommandPaletteVisibility } from '@gms/ui-state';
import * as ReactRedux from 'react-redux';
import { bindActionCreators } from 'redux';

import { CommandPaletteComponent } from './command-palette-component';
import type { CommandPaletteComponentProps } from './types';

/**
 * Mapping redux state to the properties of the component
 *
 * @param state App state, root level redux store
 */
const mapStateToProps = (state: AppState): Partial<CommandPaletteComponentProps> => ({
  commandPaletteIsVisible: state.app.common.commandPaletteIsVisible,
  keyPressActionQueue: state.app.common.keyPressActionQueue
});

/**
 * Mapping methods (actions and operations) to dispatch one or more updates to the redux store
 *
 * @param dispatch the redux dispatch event alerting the store has changed
 */
const mapDispatchToProps = (
  dispatch
): Partial<React.PropsWithChildren<CommandPaletteComponentProps>> =>
  bindActionCreators(
    {
      setCommandPaletteVisibility,
      setKeyPressActionQueue: commonActions.setKeyPressActionQueue
    },
    dispatch
  );

/**
 * A new redux component that is wrapping the CommandPalette component and injecting in the redux state
 */
export const ReduxCommandPaletteContainer = compose(
  ReactRedux.connect(mapStateToProps, mapDispatchToProps)
)(CommandPaletteComponent);
