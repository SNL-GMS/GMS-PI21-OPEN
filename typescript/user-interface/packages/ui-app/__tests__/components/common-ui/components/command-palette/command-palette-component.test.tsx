import { getStore } from '@gms/ui-state';
import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import { CommandPaletteComponent } from '../../../../../src/ts/components/common-ui/components/command-palette/command-palette-component';
import type { CommandPaletteComponentProps } from '../../../../../src/ts/components/common-ui/components/command-palette/types';

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

// eslint-disable-next-line import/no-deprecated
const lodash = jest.requireActual('lodash');
lodash.uniqueId = () => '1';

describe('Command Palette Component', () => {
  it('should be defined', () => {
    expect(CommandPaletteComponent).toBeDefined();
  });

  const props: CommandPaletteComponentProps = {
    commandPaletteIsVisible: true,
    keyPressActionQueue: undefined,
    setCommandPaletteVisibility: jest.fn(),
    setKeyPressActionQueue: jest.fn()
  };

  it('matches a snapshot', () => {
    const { container } = render(
      <Provider store={getStore()}>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <CommandPaletteComponent {...props} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
