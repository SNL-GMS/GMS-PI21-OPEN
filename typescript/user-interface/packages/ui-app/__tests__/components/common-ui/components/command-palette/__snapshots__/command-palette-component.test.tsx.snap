// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command Palette Component matches a snapshot 1`] = `<div />`;
